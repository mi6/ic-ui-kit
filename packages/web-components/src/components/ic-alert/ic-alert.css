@import "../../global/normalize.css";

:host {
  display: block;
}

.container {
  min-height: 3.5rem;
  border-radius: var(--ic-border-radius);
  position: relative;
  display: flex;
  align-items: center;
}

.container-neutral {
  background-color: var(--ic-status-unknown-light);
}

.container-info {
  background-color: var(--ic-status-info-light);
}

.container-warning {
  background-color: var(--ic-status-warning-light);
}

.container-error {
  background-color: var(--ic-status-error-light);
}

.container-success {
  background-color: var(--ic-status-success-light);
}

.alert-icon-container {
  height: 100%;
  display: flex;
  align-items: center;
}

.divider {
  height: 100%;
  width: var(--ic-space-xs);
  border-radius: var(--ic-space-xxxs) 0 0 var(--ic-space-xxxs);
  position: absolute;
}

.divider-neutral {
  background-color: var(--ic-status-unknown);
}

.divider-info {
  background-color: var(--ic-status-info);
}

.divider-warning {
  background-color: var(--ic-status-warning);
}

.divider-error {
  background-color: var(--ic-status-error);
}

.divider-success {
  background-color: var(--ic-status-success);
}

.alert-icon {
  height: var(--ic-space-lg);
  width: 1.375rem;
  margin-left: 1.125rem;
}

.alert-icon > svg {
  height: var(--ic-space-lg);
  width: var(--ic-space-lg);
  display: inline-block;
}

.icon-neutral > svg {
  fill: var(--ic-status-unknown);
}

:host([variant="info"]) .alert-icon svg {
  fill: var(--ic-status-info);
}

:host([variant="warning"]) .alert-icon svg {
  fill: var(--ic-status-warning);
}

:host([variant="error"]) .alert-icon svg {
  fill: var(--ic-status-error);
}

:host([variant="success"]) .alert-icon svg {
  fill: var(--ic-status-success);
}

.alert-content {
  display: flex;
  align-items: center;
  margin-left: 0.625rem;
  width: 100%;
}

.alert-message {
  display: flex;
  align-items: center;
  padding: var(--ic-space-xs) var(--ic-space-xs) var(--ic-space-xs) 0;
  flex: 1;
}

.alert-message-title-above {
  display: inline;
  font-size: 0;
}

.alert-title {
  margin-right: var(--ic-space-xs);
}

.alert-title-above {
  white-space: normal;
}

.alert-action-container {
  margin-right: var(--ic-space-xs);
  display: flex;
  align-items: center;
}

.dismiss-icon {
  margin-right: var(--ic-space-xxxs);
  margin-left: -0.375rem;
  padding: 0.375rem;
  border: none;
  border-radius: 50%;
  background-color: inherit;
  display: flex;
  align-items: center;
  justify-content: center;
}

.dismiss-icon:hover {
  cursor: pointer;
}

@media (max-width: 628px) {
  .alert-content {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }

  .alert-message {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
  }

  .alert-title {
    white-space: normal;
  }

  .alert-action-container {
    margin-bottom: var(--ic-space-xs);
  }
}

@media (forced-colors: active) {
  .container {
    border: var(--ic-hc-border);
  }
}
