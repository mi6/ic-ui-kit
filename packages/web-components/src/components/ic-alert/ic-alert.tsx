import {
  Component,
  Element,
  Event,
  EventEmitter,
  Host,
  State,
  Listen,
  Prop,
  h,
} from "@stencil/core";
import closeIcon from "../../assets/close-icon.svg";
import { isSlotUsed } from "../../utils/helpers";
import { IcThemeForegroundEnum, IcStatusVariants } from "../../utils/types";
import { VARIANT_ICONS } from "../../utils/constants";

/**
 * @slot message - Content is placed to the right of the title.
 * @slot action - Content is placed to the right of the message.
 */
@Component({
  tag: "ic-alert",
  styleUrl: "ic-alert.css",
  shadow: true,
})
export class Alert {
  @Element() el: HTMLIcAlertElement;

  @State() alertTitleWrap: boolean = false;
  @State() visible: boolean = true;

  /**
   * If `true`, the alert will have the 'alert' ARIA role and will be announced to screen readers.
   */
  @Prop() announced?: boolean = true;

  /**
   * If `true`, the alert will have a close icon at the end to dismiss it.
   */
  @Prop() dismissible?: boolean = false;

  /**
   * The optional title to display at the start of the alert.
   */
  @Prop() heading?: string = "";

  /**
   * The main body message of the alert.
   */
  @Prop() message?: string;

  /**
   * If `true`, the title and message will appear inline instead of above and below.
   */
  @Prop() titleAbove?: boolean = false;

  /**
   * The variant of the alert which will be rendered.
   */
  @Prop() variant?: IcStatusVariants = "neutral";

  /**
   * @deprecated This event should not be used anymore. Use icDismiss instead.
   */
  @Event() dismiss: EventEmitter<void>;

  /**
   * Is emitted when the user dismisses the alert.
   */
  @Event() icDismiss: EventEmitter<void>;

  componentDidLoad(): void {
    this.alertTitleShouldWrap();
  }

  @Listen("icDismiss", { capture: true })
  handleClick(): void {
    this.visible = !this.visible;
  }

  private dismissAction = (): void => {
    this.dismiss.emit();
    this.icDismiss.emit();
  };

  private alertTitleShouldWrap() {
    const titleHeight =
      this.el.shadowRoot.querySelector(".alert-title")?.clientHeight;
    if (titleHeight > 24) this.alertTitleWrap = true;
  }

  render() {
    const {
      variant,
      heading,
      message,
      titleAbove,
      dismissible,
      announced,
      visible,
    } = this;

    return (
      visible && (
        <Host
          role={announced && "alert"}
          class={{
            [IcThemeForegroundEnum.Dark]: true,
          }}
        >
          <div
            class={{
              ["container"]: true,
              [`container-${variant}`]: true,
            }}
          >
            <div class="alert-icon-container">
              <div
                class={{
                  ["divider"]: true,
                  [`divider-${variant}`]: true,
                }}
              ></div>
              <span
                class={{
                  ["alert-icon"]: true,
                  ["svg-container"]: true,
                  [`icon-${variant}`]: true,
                }}
                innerHTML={VARIANT_ICONS[variant].icon}
              ></span>
            </div>
            <div class="alert-content">
              <div
                class={{
                  ["alert-message"]: true,
                  ["alert-message-title-above"]:
                    titleAbove || this.alertTitleWrap,
                }}
              >
                {heading && (
                  <ic-typography
                    class={{
                      ["alert-title"]: true,
                      ["alert-title-above"]: titleAbove || this.alertTitleWrap,
                    }}
                    variant="subtitle-large"
                  >
                    <p>{heading}</p>
                  </ic-typography>
                )}
                <slot name="message">
                  <ic-typography variant="body">{message}</ic-typography>
                </slot>
              </div>
              {isSlotUsed(this.el, "action") && (
                <div class="alert-action-container">
                  <slot name="action" />
                </div>
              )}
            </div>
            <div class="dismiss-icon-container">
              {dismissible && (
                <ic-button
                  class={{
                    ["svg-container"]: true,
                    ["dismiss-icon"]: true,
                  }}
                  innerHTML={closeIcon}
                  onClick={this.dismissAction}
                  variant="icon"
                  appearance={IcThemeForegroundEnum.Dark}
                  title="dismiss"
                ></ic-button>
              )}
            </div>
          </div>
        </Host>
      )
    );
  }
}
