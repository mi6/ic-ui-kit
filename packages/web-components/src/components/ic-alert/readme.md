# ic-alert



<!-- Auto Generated Below -->


## Properties

| Property      | Attribute     | Description                                                                                   | Type                                                       | Default     |
| ------------- | ------------- | --------------------------------------------------------------------------------------------- | ---------------------------------------------------------- | ----------- |
| `announced`   | `announced`   | If `true`, the alert will have the 'alert' ARIA role and will be announced to screen readers. | `boolean`                                                  | `true`      |
| `dismissible` | `dismissible` | If `true`, the alert will have a close icon at the end to dismiss it.                         | `boolean`                                                  | `false`     |
| `heading`     | `heading`     | The optional title to display at the start of the alert.                                      | `string`                                                   | `""`        |
| `message`     | `message`     | The main body message of the alert.                                                           | `string`                                                   | `undefined` |
| `titleAbove`  | `title-above` | If `true`, the title and message will appear inline instead of above and below.               | `boolean`                                                  | `false`     |
| `variant`     | `variant`     | The variant of the alert which will be rendered.                                              | `"error" \| "info" \| "neutral" \| "success" \| "warning"` | `"neutral"` |


## Events

| Event       | Description                                                                                                             | Type                |
| ----------- | ----------------------------------------------------------------------------------------------------------------------- | ------------------- |
| `dismiss`   | <span style="color:red">**[DEPRECATED]**</span> This event should not be used anymore. Use icDismiss instead.<br/><br/> | `CustomEvent<void>` |
| `icDismiss` | Is emitted when the user dismisses the alert.                                                                           | `CustomEvent<void>` |


## Slots

| Slot        | Description                                    |
| ----------- | ---------------------------------------------- |
| `"action"`  | Content is placed to the right of the message. |
| `"message"` | Content is placed to the right of the title.   |


## Dependencies

### Used by

 - [ic-dialog](../ic-dialog)

### Depends on

- [ic-typography](../ic-typography)
- [ic-button](../ic-button)

### Graph
```mermaid
graph TD;
  ic-alert --> ic-typography
  ic-alert --> ic-button
  ic-button --> ic-loading-indicator
  ic-button --> ic-tooltip
  ic-loading-indicator --> ic-typography
  ic-tooltip --> ic-typography
  ic-dialog --> ic-alert
  style ic-alert fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


