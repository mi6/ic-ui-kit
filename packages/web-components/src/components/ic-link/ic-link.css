@import "../../global/normalize.css";

:host(.link) .ic-link,
:host(.link) ::slotted(a) {
  color: var(--ic-hyperlink);
  text-decoration: underline;
  font-weight: var(--ic-font-weight-bold);
  transition: var(--ic-easing-transition-fast);
}

:host(.link.dark) .ic-link.dark,
:host(.link.dark) ::slotted(a) {
  color: var(--ic-color-primary-text);
}

:host(.link.light) .ic-link.light,
:host(.link.light) ::slotted(a) {
  color: var(--ic-color-white-text);
}

:host(.link) .ic-link:visited,
:host(.link) ::slotted(a:visited) {
  color: var(--ic-hyperlink-visited);
}

:host(.link) .ic-link:hover,
:host(.link) .ic-link:focus,
:host(.link) ::slotted(a:hover),
:host(.link) ::slotted(a:focus) {
  outline: none;
  border-bottom: 0.25rem solid !important;
  margin-bottom: -0.25rem !important;
  text-decoration: none;
}

@supports (text-decoration-thickness: 25%) {
  :host(.link) .ic-link:hover,
  :host(.link) .ic-link:focus,
  :host(.link) ::slotted(a:hover),
  :host(.link) ::slotted(a:focus) {
    text-decoration-line: underline;
    text-decoration-thickness: 25%;
    text-underline-offset: 25%;
    border-bottom: 0 !important;
    margin-bottom: 0 !important;
  }
}

:host(.link) .ic-link:active,
:host(.link) .ic-link:focus:active,
:host(.link) .ic-link:visited:active,
:host(.link) ::slotted(a:active),
:host(.link) ::slotted(a:focus:active),
:host(.link) ::slotted(a:visited:active) {
  text-decoration: none;
}

.ic-link-open-in-new-icon {
  vertical-align: middle;
  margin-left: var(--ic-space-xxs);
}

.ic-link > .ic-link-open-in-new-icon {
  fill: var(--ic-hyperlink);
}

.ic-link.dark > .ic-link-open-in-new-icon {
  fill: var(--ic-color-primary-text);
}

.ic-link.light > .ic-link-open-in-new-icon {
  fill: var(--ic-color-white-text);
}

.ic-link-open-in-new-icon > svg {
  width: var(--ic-space-md);
  height: var(--ic-space-md);
  fill: currentcolor;
}

.ic-link:visited > .ic-link-open-in-new-icon {
  fill: var(--ic-hyperlink-visited);
}

.ic-link.light:visited > .ic-link-open-in-new-icon {
  fill: var(--ic-hyperlink-contrast-visited);
}

:host(.breadcrumb-link) .ic-link {
  display: var(--breadcrumb-link-display);
  align-items: var(--breadcrumb-link-align-items);
  gap: var(--breadcrumb-link-gap);
}

:host(.breadcrumb-link) .ic-link ::slotted(.back-icon) {
  height: var(--ic-space-lg);
  width: var(--ic-space-lg);
}

:host(.breadcrumb-link.current-page) a,
:host(.breadcrumb-link.current-page) ::slotted(a) {
  font-weight: normal;
  color: inherit;
  text-decoration: none;
  display: flex;
  align-items: center;
}

:host(.breadcrumb-link.current-page) .ic-link,
:host(.breadcrumb-link.current-page) ::slotted(a:focus) {
  outline: var(--ic-hc-focus-outline);
  text-decoration: none;
}

:host(.breadcrumb-link.current-page) .ic-link:visited {
  color: var(--ic-color-primary-text);
}
