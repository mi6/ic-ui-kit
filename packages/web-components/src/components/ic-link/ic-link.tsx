import {
  Component,
  Element,
  Prop,
  h,
  Host,
  Listen,
  Method,
} from "@stencil/core";

import OpenInNew from "../../assets/OpenInNew.svg";
import { getThemeFromContext, inheritAttributes } from "../../utils/helpers";
import { IC_INHERITED_ARIA } from "../../utils/constants";
import {
  IcTheme,
  IcThemeForeground,
  IcThemeForegroundEnum,
} from "../../utils/types";

/**
 * @slot router-item - Handle routing by nesting your routes in this slot.
 */

@Component({
  tag: "ic-link",
  styleUrl: "ic-link.css",
  shadow: {
    delegatesFocus: true,
  },
})
export class Link {
  private inheritedAttributes: { [k: string]: unknown } = {};
  private routerSlot: HTMLElement;

  @Element() el: HTMLIcLinkElement;

  /**
   * The appearance of the link, e.g. dark, light, or default.
   */
  @Prop({ mutable: true }) appearance?: IcThemeForeground = "default";

  /**
   * If `true`, the user can save the linked URL instead of navigating to it.
   */
  @Prop() download?: string | boolean = false;

  /**
   * The URL that the link points to.
   */
  @Prop() href?: string = null;

  /**
   * The human language of the linked URL.
   */
  @Prop() hreflang?: string;

  /**
   * How much of the referrer to send when following the link.
   */
  @Prop() referrerpolicy?: ReferrerPolicy;

  /**
   * The relationship of the linked URL as space-separated link types.
   */
  @Prop() rel?: string;

  /**
   * @deprecated This prop should not be used anymore. The 'open in new tab/window' icon will display automatically if target="_blank".
   */
  @Prop() showIcon?: boolean;

  /**
   * The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).
   */
  @Prop() target?: string;

  componentWillLoad(): void {
    this.inheritedAttributes = inheritAttributes(this.el, [
      ...IC_INHERITED_ARIA,
      "aria-expanded",
    ]);

    this.updateTheme();
  }

  @Listen("themeChange", { target: "document" })
  themeChangeHandler(ev: CustomEvent): void {
    const theme: IcTheme = ev.detail;
    this.updateTheme(theme.mode);
  }

  /**
   * Sets focus on the link.
   */
  @Method()
  async setFocus(): Promise<void> {
    if (this.el.shadowRoot.querySelector("a")) {
      this.el.shadowRoot.querySelector("a").focus();
    }
  }

  private updateTheme(newTheme: IcThemeForeground = null): void {
    const theme = getThemeFromContext(this.el, newTheme || null);

    switch (theme) {
      case IcThemeForegroundEnum.Light:
        this.appearance = IcThemeForegroundEnum.Light;
        break;
      case IcThemeForegroundEnum.Dark:
        this.appearance = IcThemeForegroundEnum.Dark;
        break;
    }
  }

  private hasRouterSlot(): boolean {
    this.routerSlot = this.el.querySelector('[slot="router-item"]');
    if (this.routerSlot) {
      this.routerSlot.ariaLabel = this.routerSlot.textContent;
    }
    return !!this.routerSlot;
  }

  render() {
    const {
      download,
      href,
      hreflang,
      referrerpolicy,
      rel,
      target,
      showIcon,
      appearance,
    } = this;

    return (
      <Host class={{ ["link"]: true, [`${appearance}`]: true }}>
        {this.hasRouterSlot() ? (
          <slot name="router-item"></slot>
        ) : (
          <a
            class={{
              ["ic-link"]: href !== null,
              [`${appearance}`]: href !== null,
            }}
            download={download !== false ? download : null}
            href={href}
            hrefLang={hreflang}
            referrerPolicy={referrerpolicy}
            rel={rel}
            target={target}
            tabindex={href !== null ? "0" : "-1"}
            {...this.inheritedAttributes}
          >
            <slot />
            {(showIcon || target === "_blank") && (
              <span class="ic-link-open-in-new-icon" innerHTML={OpenInNew} />
            )}
          </a>
        )}
      </Host>
    );
  }
}
