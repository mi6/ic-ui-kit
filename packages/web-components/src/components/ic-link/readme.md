# ic-link



<!-- Auto Generated Below -->


## Properties

| Property         | Attribute        | Description                                                                                                                                                                      | Type                                                                                                                                                                                     | Default     |
| ---------------- | ---------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| `appearance`     | `appearance`     | The appearance of the link, e.g. dark, light, or default.                                                                                                                        | `"dark" \| "default" \| "light"`                                                                                                                                                         | `"default"` |
| `download`       | `download`       | If `true`, the user can save the linked URL instead of navigating to it.                                                                                                         | `boolean \| string`                                                                                                                                                                      | `false`     |
| `href`           | `href`           | The URL that the link points to.                                                                                                                                                 | `string`                                                                                                                                                                                 | `null`      |
| `hreflang`       | `hreflang`       | The human language of the linked URL.                                                                                                                                            | `string`                                                                                                                                                                                 | `undefined` |
| `referrerpolicy` | `referrerpolicy` | How much of the referrer to send when following the link.                                                                                                                        | `"" \| "no-referrer" \| "no-referrer-when-downgrade" \| "origin" \| "origin-when-cross-origin" \| "same-origin" \| "strict-origin" \| "strict-origin-when-cross-origin" \| "unsafe-url"` | `undefined` |
| `rel`            | `rel`            | The relationship of the linked URL as space-separated link types.                                                                                                                | `string`                                                                                                                                                                                 | `undefined` |
| `showIcon`       | `show-icon`      | <span style="color:red">**[DEPRECATED]**</span> This prop should not be used anymore. The 'open in new tab/window' icon will display automatically if target="_blank".<br/><br/> | `boolean`                                                                                                                                                                                | `undefined` |
| `target`         | `target`         | The place to display the linked URL, as the name for a browsing context (a tab, window, or iframe).                                                                              | `string`                                                                                                                                                                                 | `undefined` |


## Methods

### `setFocus() => Promise<void>`

Sets focus on the link.

#### Returns

Type: `Promise<void>`




## Slots

| Slot            | Description                                         |
| --------------- | --------------------------------------------------- |
| `"router-item"` | Handle routing by nesting your routes in this slot. |


## Dependencies

### Used by

 - [ic-breadcrumb](../ic-breadcrumb)

### Graph
```mermaid
graph TD;
  ic-breadcrumb --> ic-link
  style ic-link fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------


