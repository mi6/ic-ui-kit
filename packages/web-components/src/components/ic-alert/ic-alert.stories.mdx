import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Description,
} from "@storybook/addon-docs";
import { html } from "lit-html";
import readme from "./readme.md";
import Link from "../ic-link/readme.md";
import Button from "../ic-button/readme.md";

<Meta title="Web Components/Alert" component="ic-alert" />

<Description markdown={readme} />

export const defaultArgs = {
  dismissible: false,
  heading: "Heading",
  message: "Message",
  titleAbove: false,
  variant: "neutral",
};

### Variants

<Canvas>
  <Story name="Variants" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert
        heading="Neutral"
        message="This alert is for displaying miscellaneous messages"
      ></ic-alert>
      <ic-alert
        variant="info"
        heading="Info"
        message="This alert is for displaying information"
      ></ic-alert>
      <ic-alert
        variant="error"
        heading="Error"
        message="This alert is for displaying errors"
      ></ic-alert>
      <ic-alert
        variant="warning"
        heading="Warning"
        message="This alert is for displaying warnings"
      ></ic-alert>
      <ic-alert
        variant="success"
        heading="Success"
        message="This alert is for displaying success messages. It has a very long message with lots of text which goes over multiple
        lines so the alert should expand as well to make sure it doesn't overflow"
      ></ic-alert>`}
  </Story>
</Canvas>

### Message only

<Canvas>
  <Story name="Message only" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert
      message="This alert is for displaying miscellaneous messages"
    ></ic-alert>`}
  </Story>
</Canvas>

### Dismissible

<Canvas>
  <Story name="Dismissible" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert
      heading="Neutral"
      message="This alert is for displaying miscellaneous messages"
      dismissible="true"
    ></ic-alert>`}
  </Story>
</Canvas>

### With action

<Canvas>
  <Story name="With action" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert
      heading="Neutral"
      message="This alert is for displaying miscellaneous messages"
      ><ic-button slot="action" variant="secondary">Button</ic-button></ic-alert
    >`}
  </Story>
</Canvas>

### Dismissible with action

<Canvas>
  <Story name="Dismissible with action" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert
      heading="Neutral"
      message="This alert is for displaying miscellaneous messages"
      dismissible="true"
      ><ic-button slot="action" variant="secondary">Button</ic-button></ic-alert
    >`}
  </Story>
</Canvas>

### Title above

<Canvas>
  <Story name="Title above" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert
      heading="Neutral"
      message="This alert is for displaying miscellaneous messages"
      title-above="true"
    ></ic-alert>`}
  </Story>
</Canvas>

### Responsive title

<Canvas>
  <Story name="Responsive title" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert
        heading="This title is very long so should force the alert to add 'title-above'"
        message="This alert does not have 'title-above' added as a prop but it is forced to add it due to the length of the title"
      ></ic-alert>
      <ic-alert
        variant="success"
        heading="Shorter title"
        message="Neither of these has 'title-above' set to true"
      ></ic-alert>`}
  </Story>
</Canvas>

### Custom message

<Canvas>
  <Story name="Custom message" parameters={{ loki: { skip: true } }}>
    {html`<ic-alert heading="This alert uses a custom message slot">
      <ic-typography variant="body" slot="message">
        <p>
          This is some text and
          <ic-link href="/" inline>this is an inline link</ic-link> within the
          text.
        </p>
      </ic-typography>
    </ic-alert>`}
  </Story>
</Canvas>

### Custom message and title above

<Canvas>
  <Story
    name="Custom message and title above"
    parameters={{ loki: { skip: true } }}
  >
    {html`<ic-alert
      heading="Want to know more about our coffee?"
      title-above="true"
    >
      <span slot="message" style="font:var(--ic-font-body)">
        Go to our <ic-link href="/">coffee page</ic-link> to learn more.
      </ic-typography>
    </ic-alert>`}
  </Story>
</Canvas>

### Announced

<Canvas>
  <Story name="Announced" parameters={{ loki: { skip: true } }}>
    {html`
      <ic-alert
        id="alert"
        heading="Neutral"
        message="This alert is for displaying miscellaneous messages"
        announced="false"
      ></ic-alert>
      <br />
      <br />
      <ic-button variant="primary" size="small" onClick="onClick()" id="btn"
        >Toggle announced prop</ic-button
      >
      <script>
        const alert = document.getElementById("alert");
        const btn = document.getElementById("btn");
        btn.addEventListener("click", onClick);
        function onClick(e) {
          e.preventDefault();
          alert.announced
            ? (alert.announced = "false")
            : (alert.announced = "true");
        }
      </script>
    `}
  </Story>
</Canvas>

### Playground

<Canvas>
  <Story
    args={defaultArgs}
    argTypes={{
      variant: {
        options: ["neutral", "info", "warning", "error", "success"],
        control: { type: "select" },
      },
    }}
    name="Playground"
    parameters={{ loki: { skip: true } }}
  >
    {(args) =>
      html`<ic-alert
        dismissible=${args.dismissible}
        variant=${args.variant}
        heading=${args.heading}
        message=${args.message}
        title-above=${args.titleAbove}
      ></ic-alert>`
    }
  </Story>
</Canvas>
