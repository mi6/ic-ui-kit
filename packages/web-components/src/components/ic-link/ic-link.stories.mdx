import {
  Meta,
  Story,
  Canvas,
  ArgsTable,
  Description,
} from "@storybook/addon-docs";
import { html } from "lit-html";
import readme from "./readme.md";

<Meta title="Web Components/Link" component="ic-link" />

<Description markdown={readme} />

### Default

<Canvas>
  <Story
    name="Default"
    parameters={{ layout: "fullscreen", loki: { skip: true } }}
  >
    {html`<div>
      <ic-link href="/">This is a link</ic-link>
      <br />
      <br />
      <div style="width: 100px">
        <ic-link href="/"
          >This is a very very long link that goes multi-line</ic-link
        >
      </div>
    </div>`}
  </Story>
</Canvas>

### With icon

<Canvas>
  <Story
    name="With icon"
    parameters={{ layout: "fullscreen", loki: { skip: true } }}
  >
    {html`<ic-link href="/" show-icon>This is a link</ic-link>`}
  </Story>
</Canvas>

### With download

<Canvas>
  <Story
    name="Download link"
    parameters={{ layout: "fullscreen", loki: { skip: true } }}
  >
    {html`<ic-link href="/" download>Download File</ic-link>`}
  </Story>
</Canvas>

### Dark

<Canvas>
  <Story
    name="Dark"
    parameters={{ layout: "fullscreen", loki: { skip: true } }}
  >
    {html`<ic-link href="/" appearance="dark" target="_blank"
      >This is a link</ic-link
    >`}
  </Story>
</Canvas>

### Light

<Canvas>
  <Story
    name="Light"
    parameters={{ backgrounds: { default: "dark" }, loki: { skip: true } }}
  >
    {html` <ic-link href="/" appearance="light" show-icon
      >This is a link</ic-link
    >`}
  </Story>
</Canvas>
