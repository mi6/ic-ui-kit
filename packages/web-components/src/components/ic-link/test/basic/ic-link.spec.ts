import { newSpecPage } from "@stencil/core/testing";
import { Alert } from "../../../ic-alert/ic-alert";
import { Link } from "../../ic-link";

describe("ic-link component", () => {
  it("should render a link by default with the correct text", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link>IC Link Test</ic-link>`,
    });

    expect(page.root).toEqualHtml(`
      <ic-link class="default link">
      <mock:shadow-root>
        <a tabindex="-1">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should apply the 'download' attribute to the anchor element", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link download="test-download">IC Link Test</ic-link>`,
    });

    expect(page.root)
      .toEqualHtml(`<ic-link class="default link" download="test-download">
    <mock:shadow-root>
        <a download="test-download" tabindex="-1">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should apply the 'href' attribute and ic-link styling to the anchor element", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link href="test-href">IC Link Test</ic-link>`,
    });

    expect(page.root)
      .toEqualHtml(`<ic-link href="test-href" class="default link">
    <mock:shadow-root>
        <a class="default ic-link" href="test-href" tabindex="0">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should apply the 'dark' appearance styling and ic-link styling", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link  href="test-href" appearance="dark">IC Link Test</ic-link>`,
    });

    expect(page.root)
      .toEqualHtml(`<ic-link href="test-href" appearance="dark" class="dark link">
    <mock:shadow-root>
        <a class="dark ic-link" href="test-href" tabindex="0">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should apply the 'hreflang' attribute to the anchor element", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link hreflang="test-hreflang">IC Link Test</ic-link>`,
    });

    expect(page.root)
      .toEqualHtml(`<ic-link class="default link" hreflang="test-hreflang">
    <mock:shadow-root>
        <a hreflang="test-hreflang" tabindex="-1">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should apply the 'referrerpolicy' attribute to the anchor element", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link referrerpolicy="no-referrer">IC Link Test</ic-link>`,
    });

    expect(page.root)
      .toEqualHtml(`<ic-link class="default link" referrerpolicy="no-referrer">
    <mock:shadow-root>
        <a referrerpolicy="no-referrer" tabindex="-1">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should apply the 'rel' attribute to the anchor element", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link rel="test-rel">IC Link Test</ic-link>`,
    });

    expect(page.root).toEqualHtml(`<ic-link class="default link" rel="test-rel">
    <mock:shadow-root>
        <a rel="test-rel" tabindex="-1">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should apply the 'target' attribute to the anchor element", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link target="test-target">IC Link Test</ic-link>`,
    });

    expect(page.root)
      .toEqualHtml(`<ic-link class="default link" target="test-target">
    <mock:shadow-root>
        <a target="test-target" tabindex="-1">
        <slot></slot>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should render an 'open in new' icon when 'show-icon' is provided", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link show-icon>IC Link Test</ic-link>`,
    });

    expect(page.root).toEqualHtml(`<ic-link class="default link" show-icon>
    <mock:shadow-root>
        <a tabindex="-1">
          <slot></slot>
          <span class="ic-link-open-in-new-icon">
            svg
          </span>
        </a>
        </mock:shadow-root>
        IC Link Test
      </ic-link>
    `);
  });

  it("should render a slotted link", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link>
        <ic-link slot="router-item">Slotted Link</ic-link>
      </ic-link>`,
    });

    expect(page.root).toEqualHtml(`<ic-link class="default link">
    <mock:shadow-root>
      <slot name="router-item"></slot>
    </mock:shadow-root>
    <ic-link class="default link" slot="router-item">
      <mock:shadow-root>
        <a tabindex="-1">
          <slot></slot>
        </a>
      </mock:shadow-root>
      Slotted Link
      </ic-link>
      </ic-link>
    `);
  });

  it("should get theme from context", async () => {
    const page = await newSpecPage({
      components: [Alert, Link],
      html: `<ic-alert message="This has an action"><ic-link slot="action">Theme context link</ic-link></ic-alert>`,
    });

    expect(page.root).toMatchSnapshot("alert-theme-context");
  });

  it("correctly sets variant on theme change", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-hero heading="Hero heading" subheading="Hero description">
      <ic-link slot="interaction">Link</ic-link>
    </ic-hero>
`,
    });

    page.rootInstance.themeChangeHandler({ detail: { mode: "dark" } });
    await page.waitForChanges();

    expect(page.rootInstance.appearance).toBe("dark");

    page.rootInstance.themeChangeHandler({ detail: { mode: "light" } });
    await page.waitForChanges();

    expect(page.rootInstance.appearance).toBe("light");
  });

  it("should call 'setFocus' when link is focused", async () => {
    const page = await newSpecPage({
      components: [Link],
      html: `<ic-link>IC Link Test</ic-link>`,
    });

    //Can't expect anything in this test - this is to increase code coverage only
    await page.rootInstance.setFocus().toHaveBeenCalled;
  });
});
