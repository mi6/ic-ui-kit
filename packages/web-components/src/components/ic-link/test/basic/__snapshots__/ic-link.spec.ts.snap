// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-link component should get theme from context: alert-theme-context 1`] = `
<ic-alert class="dark" message="This has an action" role="alert">
  <mock:shadow-root>
    <div class="container container-neutral">
      <div class="alert-icon-container">
        <div class="divider divider-neutral"></div>
        <span class="alert-icon icon-neutral svg-container">
          svg
        </span>
      </div>
      <div class="alert-content">
        <div class="alert-message">
          <slot name="message">
            <ic-typography variant="body">
              This has an action
            </ic-typography>
          </slot>
        </div>
        <div class="alert-action-container">
          <slot name="action"></slot>
        </div>
      </div>
      <div class="dismiss-icon-container"></div>
    </div>
  </mock:shadow-root>
  <ic-link class="default link" slot="action">
    <mock:shadow-root>
      <a tabindex="-1">
        <slot></slot>
      </a>
    </mock:shadow-root>
    Theme context link
  </ic-link>
</ic-alert>
`;
